internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -170,49 +194,126 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
+		return nil, fmt.Errorf("versionedhome is empty: %v", unpackRes)
 	}
 
-	if err := copyActionStore(u.log, newHash); err != nil {
+	newHome := filepath.Join(paths.Top(), unpackRes.VersionedHome)
+
+	if err := copyActionStore(u.log, newHome); err != nil {
 		return nil, errors.New(err, "failed to copy action store")
 	}
 
-	if err := copyRunDirectory(u.log, newHash); err != nil {
+	newRunPath := filepath.Join(newHome, "run")
+	oldRunPath := filepath.Join(paths.Home(), "run")
+
+	if err := copyRunDirectory(u.log, oldRunPath, newRunPath); err != nil {
 		return nil, errors.New(err, "failed to copy run directory")
 	}
 
 	det.SetState(details.StateReplacing)
 
-	if err := ChangeSymlink(ctx, u.log, newHash); err != nil {
+	// create symlink to the <new versioned-home>/elastic-agent
+	hashedDir := unpackRes.VersionedHome
+
+	symlinkPath := filepath.Join(paths.Top(), agentName)
+
+	// paths.BinaryPath properly derives the binary directory depending on the platform. The path to the binary for macOS is inside of the app bundle.
+	newPath := paths.BinaryPath(filepath.Join(paths.Top(), hashedDir), agentName)
+
+	currentVersionedHome, err := filepath.Rel(paths.Top(), paths.Home())
+	if err != nil {
+		return nil, fmt.Errorf("calculating home path relative to top, home: %q top: %q : %w", paths.Home(), paths.Top(), err)
+	}
+
+	if err := ChangeSymlink(u.log, paths.Top(), symlinkPath, newPath); err != nil {
 		u.log.Errorw("Rolling back: changing symlink failed", "error.message", err)
-		rollbackInstall(ctx, u.log, newHash)
+		rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
 		return nil, err
 	}
 
-	if err := u.markUpgrade(ctx, u.log, newHash, action, det); err != nil {
+	current := agentInstall{
+		version:       version,
+		hash:          unpackRes.Hash,
+		versionedHome: unpackRes.VersionedHome,
+	}
+
+	previous := agentInstall{
+		version:       release.VersionWithSnapshot(),
+		hash:          release.Commit(),
+		versionedHome: currentVersionedHome,
+	}
+
+	if err := markUpgrade(u.log,
+		paths.Data(), //data dir to place the marker in
+		current,      //new agent version data
+		previous,     // old agent version data
+		action, det); err != nil {
 		u.log.Errorw("Rolling back: marking upgrade failed", "error.message", err)
-		rollbackInstall(ctx, u.log, newHash)
+		rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
 		return nil, err
 	}
 
-	if err := InvokeWatcher(u.log); err != nil {
+	minParsedVersionForNewUpdateMarker := agtversion.NewParsedSemVer(8, 13, 0, "", "")
+	var watcherExecutable string
+	if parsedVersion.Less(*minParsedVersionForNewUpdateMarker) {
+		// use the current agent executable for watch
+		watcherExecutable = paths.BinaryPath(paths.VersionedHome(paths.Top()), agentName)
+	} else {
+		// use the new agent executable as it should be able to parse the new update marker
+		watcherExecutable = paths.BinaryPath(filepath.Join(paths.Top(), unpackRes.VersionedHome), agentName)
+	}
```

Comment:
What exactly is it in the upgrade marker that breaks backwards compatibility?

We AFAIK didn't have this problem when we introduced upgrade details.

Or is the problem not the upgrade marker, is it that the older agent doesn't understand the path structure of the current 8.13.0 agent? If this is the case I think the comment should clarify this, and I'm unsure if there's anything else we could do about this.

Reply:
P.S. now that I've actually looked at this snippet I see how it'd be hard to pull out on its own, because it's pointless by itself.

Reply:
The upgrade marker in itself is backward compatible (I only added the fields with version and paths of new and old versions of agent).
The issue is that a watcher that does not read those fields will break the agent installation that uses the new paths on Cleanup/Rollback ( it will remove any agent directory that uses a path different from `elastic-agent-<hash>`).

The reason why we didn't have this problem for upgrade details is: those are just informational states that we send along with our status, if those are not there we simply don't set the status and the upgrade completes normally. In this case the watcher will actively delete files and directories and if it does it the legacy way on installations that use the new paths the agent is broken.